### PR TITLE
Custom cypher scalar field filtering

### PR DESCRIPTION
**Pull request by mjfwebb**
# Description

This PR adds the ability to filter on scalar custom cypher fields. A contrived example:

Type definitions:
```gql
type Thing {
    name: String
    custom_count: Int
        @cypher(
            statement: """
            MATCH (m:${CustomType})
            RETURN count(m) as c
            """
            columnName: "c"
        )
}
```

You can now query with a filter based on the `custom_count` field:
```gql
query {
    things(where: { custom_count_GT: 0 }) {
        name
        custom_count
    }
}
```

## Complexity

Medium

# Issue

Relates to #554, but not fully closing as this does not include list/relationship fields.

Closes 

# Checklist

The following requirements should have been met (depending on the changes in the branch):

- [ ] Documentation has been updated
- [x] TCK tests have been updated
- [x] Integration tests have been updated
- [ ] Example applications have been updated
- [x] New files have copyright header
- [x] CLA (https://neo4j.com/developer/cla/) has been signed

### CODE REVIEW

**Comment by neo4j-team-graphql** [comment]
# Performance Report

No Performance Changes


<details>
    <summary>Show Full Table</summary>
    
| name | dbHits | old dbHits | time (ms) | old time (ms) | maxRows |
| ---- | ------ | ---------- | --------- | ------------- | ------- |
|  aggregations.TopLevelAggregate | 3404 | 3404 | 37 | 43 | 1134 |
|  aggregations.TopLevelAggregateWithMultipleFields | 6802 | 6802 | 52 | 67 | 1134 |
|  aggregations.NestedAggregation | 15407 | 15407 | 56 | 89 | 2174 |
|  aggregations.AggregationWithWhere | 10833 | 10833 | 40 | 55 | 2174 |
|  aggregations.AggregationWhereWithinNestedRelationships | 20097917 | 20097917 | 2060 | 2194 | 2008534 |
|  aggregations.AggregationWhereWithinNestedConnections | 20097917 | 20097917 | 2131 | 2201 | 2008534 |
|  aggregations.NestedCountFromMovieToActors | 8603 | 8603 | 41 | 44 | 2174 |
|  aggregations.NestedCountFromActorsToMovie | 8791 | 8791 | 37 | 41 | 2174 |
|  aggregations.DeeplyNestedCount | 10052335 | 10052335 | 2806 | 3040 | 2008534 |
|  aggregations.InterfacesAggregations | 6242 | 6242 | 44 | 51 | 2080 |
|  aggregations.InterfacesAggregationsWithTwoFields | 11444 | 11444 | 82 | 90 | 2080 |
|  batch-create.BatchCreate | 4200 | 4200 | 124 | 160 | 600 |
|  batch-create.BatchCreateSmall | 77 | 77 | 53 | 66 | 11 |
|  connect.createAndConnect | 6433 | 6433 | 129 | 171 | 3003 |
|  connections.Connection | 12951 | 12951 | 70 | 83 | 2174 |
|  connections.NestedConnection | 37705 | 37705 | 244 | 138 | 4516 |
|  create.SimpleMutation | 7 | 7 | 70 | 74 | 1 |
|  cypher-directive.TopLevelMutationDirective | 1135 | 1135 | 27 | 28 | 1134 |
|  delete.SimpleDelete | 19401 | 19401 | 657 | 653 | 1040 |
|  delete.NestedDeleteInUpdate | 16844 | 16844 | 125 | 185 | 2040 |
|  2871.NestedRelationshipFilter | 19735 | 19632 | 61 | 68 | 4428 |
|  2925.SingleRelationshipFilter | 5245 | 5245 | 39 | 53 | 1040 |
|  2925.NestedSingleRelationshipFilter | 17641 | 17641 | 71 | 92 | 2174 |
|  2925.SingleRelationshipRequiredFilter | 5201 | 5201 | 38 | 43 | 1040 |
|  2925.NestedSingleRelationshipRequiredFilter | 9361 | 9361 | 56 | 67 | 1040 |
|  query.SimpleQuery | 3121 | 3121 | 22 | 25 | 1040 |
|  query.SimpleQueryWithRelationship | 15031 | 15031 | 43 | 40 | 2174 |
|  query.QueryWhere | 8564 | 8564 | 37 | 51 | 2154 |
|  query.SimpleQueryWithNestedWhere | 8713 | 8713 | 48 | 59 | 2154 |
|  query.Nested | 10084891 | 10084891 | 7134 | 7069 | 2008534 |
|  query.NestedWithFilter | 10064992 | 10064992 | 7208 | 6586 | 2004000 |
|  query.OrFilterOnRelationships | 37420 | 36551 | 148 | 191 | 2171 |
|  query.OrFilterOnRelationshipsAndNested | 27505 | 26982 | 194 | 214 | 2171 |
|  query.QueryWithNestedIn | 14948 | 14216 | 48 | 71 | 2116 |
|  query.NestedConnectionWhere | 8703 | 8703 | 51 | 60 | 2174 |
|  query.DeeplyNestedConnectionWhere | 8793 | 8702 | 69 | 86 | 2174 |
|  query.DeeplyNestedWithRelationshipFilters | 17489 | 17357 | 142 | 162 | 1596 |
|  query.NestedWithRelationshipSingleFilters | 3808 | 3808 | 143 | 174 | 1134 |
|  query.Fulltext | 64 | 64 | 32 | 35 | 16 |
|  query.FulltextWithNestedQuery | 516 | 516 | 42 | 48 | 84 |
|  sorting-and-cypher.TopLevelSortWithCypher | 12961 | 12961 | 39 | 45 | 2174 |
|  sorting-and-cypher.TopLevelConnectionSortWithCypher | 12961 | 12961 | 58 | 70 | 2174 |
|  sorting-and-cypher.TopLevelSortWithCypherWithNested | 13096 | 13096 | 56 | 61 | 2174 |
|  sorting-and-cypher.TopLevelConnectionSortWithCypherWithNested | 13096 | 13096 | 94 | 104 | 2174 |
|  sorting-and-cypher.TopLevelSortWithExpensiveCypher | 13658 | 13705 | 93 | 120 | 2174 |
|  sorting-and-cypher.TopLevelConnectionSortWithExpensiveCypher | 13266 | 13266 | 96 | 124 | 2174 |
|  sorting.SortMultipleTypes | 3436 | 3436 | 78 | 107 | 1040 |
|  sorting.SortMultipleTypesWithCypherWithCypher | 13321 | 13321 | 99 | 128 | 2174 |
|  sorting.SortOnNestedFields | 12951 | 12951 | 45 | 66 | 2174 |
|  sorting.SortDeeplyNestedFields | 39785 | 39785 | 81 | 133 | 4516 |
|  sorting.ConnectionWithSort | 3271 | 3271 | 70 | 77 | 1040 |
|  unions.SimpleUnionQuery | 321 | 321 | 52 | 60 | 35 |
|  unions.SimpleUnionQueryWithMissingFields | 293 | 293 | 51 | 58 | 35 |
|  unions.NestedUnion | 309975 | 309975 | 275 | 307 | 33033 |
|  unions.NestedUnionWithMissingFields | 283949 | 283949 | 251 | 254 | 33033 |
|  update.NestedUpdate | 14137 | 14137 | 107 | 108 | 2002 |

</details>
        
Old Schema Generation: 28.733s
Schema Generation: 28.772s
Old Subgraph Schema Generation: 30.869s
Subgraph Schema Generation: 30.148s

File: packages/graphql/src/translate/queryAST/ast/filters/property-filters/CypherFilter.ts [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+        return operation;
+
+        // let retProj;
+
+        // if (this.isNested && this.cypherAttributeField.typeHelper.isList()) {
+        //     retProj = [Cypher.collect(nestedContext.returnVariable), context.returnVariable];
+        // } else {
+        //     retProj = [nestedContext.returnVariable, context.returnVariable];
+        // }
+        // const ret = new Cypher.Return(retProj);
```
Comment (angrykoala) [inline packages/graphql/src/translate/queryAST/ast/filters/property-filters/CypherFilter.ts]: I would remove this commented code, maybe keep a mention to the scope, as that is an optimisation that we may want to do in the future

File: packages/graphql/tests/integration/directives/cypher/cypher-filtering.int.test.ts [hunk 2]
```diff
@@ -0,0 +1,1092 @@
...
+                        {
+                            name: "Keanu Reeves",
+                        },
+                    ],
+                },
+            ],
+        });
+    });
+
+    test("Connect filter", async () => {
```
Comment (angrykoala) [inline packages/graphql/tests/integration/directives/cypher/cypher-filtering.int.test.ts]: *Nitpick* Just for organization purposes, would it be worth splitting this test file into separate files, maybe when tackling complex cases?

Is just that it is hard to track down what features are covered and which aren't in a single test

**Review by angrykoala — DISMISSED** [review]

File: packages/graphql/src/translate/queryAST/ast/filters/property-filters/CypherFilter.ts [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+        property,
+        param,
+    }: {
+        operator: FilterOperator;
+        property: Cypher.Expr;
+        param: Cypher.Expr;
+    }): Cypher.ComparisonOp {
+        const coalesceProperty = this.coalesceValueIfNeeded(property);
+
+        if (operator === "DISTANCE") {
```
Comment (MacondoExpress) [inline packages/graphql/src/translate/queryAST/ast/filters/property-filters/CypherFilter.ts]: From what I see, I believe here there is a bug in the implementation, it uses the `createPointOperation` depending on the operator, while it should use `createPointOperation` depending on the AttributeType IMO. I believe that the same error will be present also for Duration Filters or any other filters that generate the predicate differently from PropertyFilter.

**Review by MacondoExpress — APPROVED** [review]

**Pull request merged**